src/Build/Globbing/MSBuildGlob.cs

```diff
@@ -0,0 +1,259 @@
+            FileMatcher.GetRegexMatchInfo(
+                normalizedInput,
+                _regex.Value,
+                out isMatch,
+                out fixedDirectoryPart,
+                out wildcardDirectoryPart,
+                out filenamePart);
+
+            return new MatchInfoResult(isMatch, fixedDirectoryPart, wildcardDirectoryPart, filenamePart);
+        }
+
+        private string NormalizeMatchInput(string stringToMatch)
+        {
+            var rootedInput = Path.Combine(_globRoot, stringToMatch);
+            var normalizedInput = FileUtilities.GetFullPathNoThrow(rootedInput);
+
+            // Degenerate case when the string to match is empty.
+            // Ensure trailing slash because the fixed directory part has a trailing slash.
+            if (stringToMatch == string.Empty)
+            {
+                normalizedInput = normalizedInput + Path.DirectorySeparatorChar;
+            }
+
+            return normalizedInput;
+        }
+
+        /// <summary>
+        ///     Parse the given <paramref name="fileSpec" /> into a <see cref="MSBuildGlob" /> using a given
+        ///     <paramref name="globRoot" />.
+        /// </summary>
+        /// <param name="globRoot">
+        ///     The root of the glob.
+        ///     The fixed directory part of the glob and the match arguments (<see cref="IsMatch" /> and <see cref="MatchInfo" />)
+        ///     will get normalized against this root.
+        ///     If empty, the current working directory is used.
+        ///     Cannot be null, and cannot contain invalid path arguments.
+        /// </param>
+        /// <param name="fileSpec">The string to parse</param>
+        /// <returns></returns>
+        public static MSBuildGlob Parse(string globRoot, string fileSpec)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(globRoot, nameof(globRoot));
+            ErrorUtilities.VerifyThrowArgumentNull(fileSpec, nameof(fileSpec));
+            ErrorUtilities.VerifyThrowArgumentInvalidPath(globRoot, nameof(globRoot));
+
+            if (globRoot == string.Empty)
+            {
+                globRoot = Directory.GetCurrentDirectory();
+            }
+
+            globRoot = FileUtilities.NormalizePath(globRoot).WithTrailingSlash();
+
+            string fixedDirectoryPart = null;
+            string wildcardDirectoryPart = null;
+            string filenamePart = null;
+
+            string matchFileExpression;
+            bool needsRecursion;
+            bool isLegalFileSpec;
+
+            FileMatcher.GetFileSpecInfo(
+                fileSpec,
+                out fixedDirectoryPart,
+                out wildcardDirectoryPart,
+                out filenamePart,
+                out matchFileExpression,
+                out needsRecursion,
+                out isLegalFileSpec,
+                FileMatcher.s_defaultGetFileSystemEntries,
+                (fixedDirPart, wildcardDirPart, filePart) =>
```

Comment:
Can you promote this lambda to a named function, so the name can provide some documentation as to its purpose?